### PR TITLE
Add IColumnModification and IModificationCommand to allow the implementations to be replaced easily

### LINKED ISSUE #12169: AppendInsertOperation and output parameters
EF Core version: dev

My implementation of AppendInsertOperation generates following SQL text:

INSERT INTO "DECI" ("D_6_1") VALUES (:p0) RETURNING "TEST_ID" INTO :p1;

Where
"p0" - (input) parameter from writeOperations
"p1" - (output) parameter from readOperations

```
 public override ResultSetMapping AppendInsertOperation
                                           (StringBuilder       commandStringBuilder,
                                            ModificationCommand command,
                                            int                 commandPosition)
 {
  var operations       = command.ColumnModifications;
  var writeOperations  = operations.Where(o => o.IsWrite).ToList();
  var readOperations   = operations.Where(o => o.IsRead).ToList();
  
  // ............

  return ResultSetMapping.NoResultSet;
 }//AppendInsertOperation
```

My data provider returns values of OUT-parameters through parameters object, not through DataReader.

So, For this query ExecuteReader returns "empty" DataReader, without columns.

EntityFrameworkCore returns the error

> Message: System.InvalidOperationException : The property 'TEST_ID' on entity type 'TEST_RECORD' has a temporary value while attempting to change the entity's state to 'Unchanged'. Either set a permanent value explicitly or ensure that the database is configured to generate values for this property.

Question: In theory, it is possible override EFCore classes for support OUT-parameters?

### LINKED ISSUE #17946: Refactor ColumnModification
- `UseCurrentValueParameter` , `UseOriginalValueParameter` ->  `UseCurrentValue`, `UseOriginalValue`, `UseParameter` (`Entry != null`)
- Review `Value` implementation
- Review usage in `UpdateSqlGenerator`, `MigrationsModelDiffer` and data operations

### LINKED ISSUE #23027: ColumnModification.ParameterName for OUTPUT parameter [readValue]
Hello,

Reference question: #12169

EF Core version: master
Database provider: my own
Target framework: (e.g. .NET 5.0)

I am trying to restore my project for EFCore, but I have problems :)

My implementation of UpdateSqlGenerator generates INSERT (and UPDATE) statement with RETURNING clause.

For example:
`INSERT INTO "TEST_MODIFY_ROW_WD" ("COL_TEXT_BLOB") VALUES (:p0) RETURNING "TEST_ID" INTO :p1;
`

My UpdateSqlGenerator obtains two ColumnModification objects:

1. For "COL_TEXT_BLOB" [IsWrite=true, IsRead=false, _useParameters=true]
2. For "TEST_ID" [IsWrite=false, IsRead=true, _useParameters=true]

ColumnModification.ParameterName for "COL_TEXT_BLOB" returns generated name.

ColumnModification.ParameterName for "TEST_ID" returns **null**.

Current implementation of ColumnModification.ParameterName:
https://github.com/dotnet/efcore/blob/82232b3164e8336f76a96a625eb5b0476f3c6f1e/src/EFCore.Relational/Update/ColumnModification.cs#L294-L304

Old implementation code:
```
        public virtual string ParameterName
            => _parameterName ?? (_parameterName = _generateParameterName());
```

ColumnModification.ParameterName is virtual, but I not see the way for override this method.

Could you pointed me - how I can resolve this problem?

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Fixes #23027
Fixes #12169
Fixes #17946

Co-authored-by: Kovalenko Dmitry <dmitry.lipetsk@gmail.com>

### CODE REVIEW

**Comment by dmitry-lipetsk** [comment]
Hello @AndriySvyryd

Tested. Is ok. Thanks.

File: src/EFCore.Relational/Update/IColumnModification.cs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        public bool UseCurrentValueParameter
+            => UseParameter && UseCurrentValue;
+
+        /// <summary>
+        ///     Indicates whether the original value of the property should be used
+        /// </summary>
+        public bool UseOriginalValue { get; }
+
+        /// <summary>
+        ///     Indicates whether the current value of the property should be used
```
Comment (smitpatel) [inline src/EFCore.Relational/Update/IColumnModification.cs:81]: Shouldn't all this end with a period?

**Review by smitpatel — APPROVED** [review]

**Comment by dmitry-lipetsk** [comment]
Hello EFCore team,

Provider, uses this feature, was released
https://github.com/ibprovider/Lcpi.EFCore.LcpiOleDb-v06.00

Thanks for IColumnModification and the help with other issues.

Отдельное спасибо Андрею.

Dmitry Kovalenko.

**Comment by P9avel** [comment]
Can you to show example how can i to replace IModificationCommand in my dbcontext?

**Pull request merged**